lib/longtail_platform.c

```diff
@@ -2198,14 +2218,25 @@ int Longtail_GetFilePermissions(const char* path, uint16_t* out_permissions)
     return res;
 }
 
+// Write/read in 4 MB chunks (__APPLE__ in particular will throw errno 22 if part_size is too large).
+#define LONGTAIL_PART_SIZE(cur, end) \
```

Comment:
Wondering if 4MB is good guess? Seems far from the problematic 2Gb limit.
Might want to go to 1GB or 512MB at least?
Or have you detected performance improvements with smaller write chunks?

Reply:
Definitely a guess!

Some quick non-scientific (!) tests on my Windows machine (64 GB RAM, 16 cores, current gen NVMe SSD):

| CHUNK_SIZE_MAX | PlatformWriteLargeFile | PlatformReadLargeFile |
|----------------|-----------|-----------|
| 1 MB           | 2.84916 s | 1.65695 s |
| 4 MB           | 1.98485 s | 1.70406 s |
| 16 MB          | 2.42341 s | 1.68393 s |
| 32 MB          | 3.01173 s | 1.68185 s |
| 64 MB          | 3.73048 s | 1.68759 s |
| 128 MB         | 3.73547 s | 1.68537 s |
| 512 MB         | 4.31509 s | 1.64977 s |
| DWORD_MAX      | 4.01439 s | 1.62759 s |

I don't have the facilities to do perf tests for MacOS or Linux in a real capacity (outside GitHub runners).

I think the real answer is likely "It Depends" -- operating system, disk type, driver, cache sizes etc. Finding the optimal max chunk size is probably a PR in its own right.

Are you comfortable using 4 MB as a starting point based on the above?

Reply:
Feels off to use up to 4Gb on windows but 4Mb on other platforms?
Limit to 512Mb on all platforms?

Reply:
Looking over the numbers, 4Mb on write seems faster in your test, but 512 Mb in read is faster...
Not sure what to do here...
Need a bigger set of sample data to do a proper guess.
We can have different limits for read and write for sure.

Reply:
So quick test on my machine:
```
4Mb
Longtail_Write: 4.76223 s
Longtail_Read: 2.25163 s

8Mb
Longtail_Write: 4.22588 s
Longtail_Read: 2.32714 s

16Mb
Longtail_Write: 4.00197 s
Longtail_Read: 2.3902 s

32Mb
Longtail_Write: 4.32685 s
Longtail_Read: 2.87161 s

64Mb
Longtail_Write: 4.6808 s
Longtail_Read: 3.074 s

128Mb
Longtail_Write: 4.26141 s
Longtail_Read: 3.31375 s

256Mb
Longtail_Write: 4.3902 s
Longtail_Read: 3.56751 s

512Mb
Longtail_Write: 4.6452 s
Longtail_Read: 3.78527 s

```
So 4Mb if quickest for reads, 16Mb for writes.

The penalty for going to 16Mb for reads is smaller than the penalty for going with 4Mb for writes.

So I'm going to make a very grounded descision (lol) and say: 16Mb is a good chunk size :)

Reply:
16 MB it is!

I guess a typical longtail-chunk will also fit inside a single read or write call then? (I'm not too familiar with how that chunk size is selected)